Add coverage for MSTG-CODE-9 (#2089)
OWASP/owasp-mstg: Document/0x05i-Testing-Code-Quality-and-Build-Settings.md

### Overview

The tests used to detect the presence of binary protection mechanisms heavily depend on the language used for developing the application. These are some basic guidelines:

> - In general all binaries must be tested, this includes both the app main executable as well as all libraries/dependencies.

Reviewer: ```suggestion
- In general all binaries must be tested, which includes both the main app executable as well as all libraries/dependencies.
```

> - In some cases the nature of the language they were written will offer sufficient protection (e.g. Java has Garbage Collection similar to ARC on iOS), so even if a feature is not enabled the test may pass.

Reviewer: ```suggestion
- Some languages are inherently protected against certain vulnerability classes. For example Java has Garbage Collection similar to ARC on iOS), so even if a feature is not enabled the test may pass.
```
I rewrote this a bit, but would actually like to remove it. "if a feature is not enabled the test may pass" is weird. The features are just different for those languages. If there is no risk of a memory leak, there are no 'free security features related to memory leaks' for that language so it just doesn't apply.

I think it makes sense to just focus on native code and skip the java stuff.

- **SSP** (Stack Smashing Protection aka. canaries): Helps prevent buffer overflow attacks by means of having a small integer right before the return pointer. A buffer overflow attack often overwrites a region of memory in order to overwrite the return pointer and take over the process-control. In that case, the canary gets overwritten as well. Therefore, the value of the canary is always checked to make sure it has not changed before a routine uses the return pointer on the stack.
  - It must be enabled for native code (keywords: C/C++/JNI/NDK) but it might be difficult to fully determine it.
    - NDK libraries should have it enabled since the compiler does it by default.
    - Other custom C/C++ libraries might not have it enabled.
  - If the language is considered memory safe (at least for mitigating buffer overflows), it’s ok if it’s not enabled for that specific binary.
  - Flutter will not compile using stack canaries because of the way Dart mitigates buffer overflows.

### Static Analysis

Test the app native libraries to determine if they have the PIE an SSP protections enabled.

> You can use rabin2 to get the binary information. We'll use [r2pay-v1.0.apk](https://github.com/OWASP/owasp-mstg/blob/master/Crackmes/Android/Level_04/r2pay-v1.0.apk):

Reviewer: ```suggestion
You can use rabin2 to get the binary information. We'll use [r2pay-v1.0.apk](https://github.com/OWASP/owasp-mstg/blob/master/Crackmes/Android/Level_04/r2pay-v1.0.apk) as an example:
```

> First we take a look at the main binary for the sake of clarity. You'll see that `canary` and `pic` seem to not enabled (`false`). However, they are actually not applicable, that's why the tool shows them as `false`.

Reviewer: ```suggestion
First, take a look at the main binary. You'll see that `canary` and `pic` are not enabled (`false`). However, they are actually not applicable, which is why the tool shows them as `false`.
```
Here too, if we start the entire testcase with saying "Only test native binaries", this part can become clearer.

```sh
rabin2 -I classes.dex  | grep -E "canary|pic"
canary   false
pic      false
```

What you should definitely test are the native libraries. They must have `canary` and `pic` both set to `true`.

That's the case for `libnative-lib.so`:

```sh
rabin2 -I lib/x86_64/libnative-lib.so | grep -E "canary|pic"
canary   true
pic      true
```

But not for `libtool-checker.so`:

```sh
rabin2 -I lib/x86_64/libtool-checker.so | grep -E "canary|pic"
canary   false
pic      true
```

> In this example, `libtool-checker.so` must be fixed to enable SSP.

Reviewer: ```suggestion
In this example, `libtool-checker.so` must be recompiled with SSP support.
```
'fixed' seems so dirty
